tutor/src/components/book-menu/ux.js

```diff
@@ -0,0 +1,80 @@
+import { observable, action, observe } from 'mobx';
+import scrollIntoView from 'scroll-into-view';
+
+export default class MenuUX {
+
+  @observable currentPage;
+  expanded = observable.map();
+
+  constructor({ book, currentPage, wrapper }) {
+    this.book = book;
+    this.wrapper = wrapper;
+    this.currentPage = currentPage;
+    observe(this, 'currentPage', this.onPageChange, true);
+  }
+
+  expand(page, options = { wasSelected: true }) {
+    this.eachParent(page, (node) => {
+      const expandedOptions = this.expanded.get(node.id);
+      if (!expandedOptions || !expandedOptions.wasSelected) {
+        this.expanded.set(node.id, options);
+      }
+    });
+  }
+
+  collapse(page) {
+    this.expanded.delete(page.id);
+  }
+
+  areSiblings(a, b) {
+    return a.parent == b.parent;
+  }
+
+  eachParent(node, fn) {
+    if (node) fn(node);
+    while ((node = node.parent)) {
+      fn(node);
+    }
+  }
+
+  isExpanded(node) {
+    return Boolean(this.expanded.get(node.id));
```

Comment:
This might get slow if too many things are expanded, but I'm not sure if it's an issue.

Reply:
probably not, expanded is a Map so it should be a relatively fast hash lookup